File: app/src/main/java/danb/speedrunbrowser/holders/WatchRunViewHolder.java
package danb.speedrunbrowser.holders;

import android.content.Context;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;
import danb.speedrunbrowser.R;
import danb.speedrunbrowser.api.objects.Game;
import danb.speedrunbrowser.api.objects.LeaderboardRunEntry;
import danb.speedrunbrowser.utils.ImageLoader;
import danb.speedrunbrowser.utils.ImageViewPlacerConsumer;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public class WatchRunViewHolder extends RecyclerView.ViewHolder {
    private static final String TAG = WatchRunViewHolder.class.getSimpleName();

    private RunViewHolder mLeaderboardHolder;

    private TextView mGameName;
    private ImageView mPlayerImage;
    private ImageView mGameImage;

    public WatchRunViewHolder(View v) {
        super(v);

        mLeaderboardHolder = new RunViewHolder(v);

        mGameName = v.findViewById(R.id.txtGameName);
        mGameImage = v.findViewById(R.id.imgGameIcon);
    }

    public void apply(Context context, CompositeDisposable disposables, Game game, LeaderboardRunEntry entry) {

        mLeaderboardHolder.apply(context, disposables, game, entry);

        if(entry.getRun().getGame().getNames() != null) {
            String gameAndCategoryText = entry.getRun().getGame().getNames().get("international");

            gameAndCategoryText += " \u2022 " + entry.getRun().getCategory().getName();

            if(entry.getRun().getLevel() != null)
                gameAndCategoryText += " \u2022 " + entry.getRun().getLevel().getName();

            mGameName.setText(gameAndCategoryText);
        }

        if(entry.getRun().getGame() != null && entry.getRun().getGame().getAssets() != null && entry.getRun().getGame().getAssets().getCoverLarge() != null) {
            mGameImage.setVisibility(View.VISIBLE);
            disposables.add(new ImageLoader(context).loadImage(entry.getRun().getGame().getAssets().getCoverLarge().getUri())
                    .subscribe(new ImageViewPlacerConsumer(mGameImage)));
        }
        else
            mGameImage.setVisibility(View.GONE);
    }
}

File: app/src/main/java/danb/speedrunbrowser/holders/WatchRunViewHolder.kt
package danb.speedrunbrowser.holders

import android.content.Context
import android.view.View
import android.widget.ImageView
import android.widget.TextView

import androidx.recyclerview.widget.RecyclerView
import danb.speedrunbrowser.R
import danb.speedrunbrowser.api.objects.Game
import danb.speedrunbrowser.api.objects.LeaderboardRunEntry
import danb.speedrunbrowser.utils.ImageLoader
import danb.speedrunbrowser.utils.ImageViewPlacerConsumer
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

class WatchRunViewHolder(v: View) : RecyclerView.ViewHolder(v) {

    private val mLeaderboardHolder: RunViewHolder = RunViewHolder(v)

    private val mGameName: TextView = v.findViewById(R.id.txtGameName)
    private val mGameImage: ImageView = v.findViewById(R.id.imgGameIcon)

    fun apply(context: Context, disposables: CompositeDisposable, game: Game, entry: LeaderboardRunEntry) {

        mLeaderboardHolder.apply(context, disposables, game, entry)

        var gameAndCategoryText = entry.run.game?.resolvedName ?: ""

        if(entry.run.category != null)
            gameAndCategoryText += " \u2022 " + entry.run.category.name

        if (entry.run.level?.name != null)
            gameAndCategoryText += " \u2022 " + entry.run.level.name

        mGameName.text = gameAndCategoryText

        if (entry.run.game != null && entry.run.game.assets.coverLarge != null) {
            mGameImage.visibility = View.VISIBLE

            val consumer = ImageViewPlacerConsumer(mGameImage)
            consumer.roundedCorners = context.resources.getDimensionPixelSize(R.dimen.game_cover_rounded_corners).toFloat()

            disposables.add(ImageLoader(context).loadImage(entry.run.game.assets.coverLarge.uri)
                    .subscribeOn(Schedulers.io())
                    .subscribe(consumer))
        } else
            mGameImage.visibility = View.GONE
    }

    companion object {
        private val TAG = WatchRunViewHolder::class.java.simpleName
    }
}
